locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/JcloudsLocationSecurityGroupEditor.java

```diff
@@ -0,0 +1,303 @@
+        LOG.debug("Adding permission to security group {}: {}", group.getName(), permission);
+        Callable<SecurityGroup> callable = new Callable<SecurityGroup>() {
+            @Override
+            public SecurityGroup call() throws Exception {
+                try {
+                    return securityApi.get().addIpPermission(permission, group);
+                } catch (Exception e) {
+                    Exceptions.propagateIfFatal(e);
+
+                    if (isDuplicate(e)) {
+                        return group;
+                    }
+
+                    throw Exceptions.propagate(e);
+                }
+            }
+        };
+        return runOperationWithRetry(callable);
+    }
+
+    @Deprecated // TODO improve this - shouldn't have AWS specifics in here
+    private boolean isDuplicate(Exception e) {
+        // Sometimes AWSResponseException is wrapped in an IllegalStateException
+        AWSResponseException cause = Exceptions.getFirstThrowableOfType(e, AWSResponseException.class);
+        if (cause != null) {
+            if ("InvalidPermission.Duplicate".equals(cause.getError().getCode())) {
+                return true;
+            }
+        }
+
+        if (e.toString().contains("already exists")) {
+            return true;
+        }
+
+        return false;
+    }
+
+
+    public SecurityGroup removePermission(final SecurityGroup group, final IpPermission permission) {
+        LOG.debug("Removing permission from security group {}: {}", group.getName(), permission);
+        Callable<SecurityGroup> callable = new Callable<SecurityGroup>() {
+            @Override
+            public SecurityGroup call() throws Exception {
+                return securityApi.get().removeIpPermission(permission, group);
+            }
+        };
+        return runOperationWithRetry(callable);
+    }
+
+    public SecurityGroup removePermissions(SecurityGroup group, final Iterable<IpPermission> permissions) {
+        for (IpPermission permission : permissions) {
+            group = removePermission(group, permission);
+        }
+        return group;
+    }
+
+    /**
+     * Runs the given callable. Repeats until the operation succeeds or {@link #isExceptionRetryable} indicates
+     * that the request cannot be retried.
+     */
+    protected <T> T runOperationWithRetry(Callable<T> operation) {
+        int backoff = 64;
+        Exception lastException = null;
+        for (int retries = 0; retries < 100; retries++) { // TODO this will try for about 2.0e+21 years; maybe not try so hard?
+            try {
+                return operation.call();
+            } catch (Exception e) {
+                lastException = e;
+                if (isExceptionRetryable.apply(e)) {
+                    LOG.debug("Attempt #{} failed to add security group: {}", retries + 1, e.getMessage());
+                    try {
+                        Thread.sleep(backoff);
```

Comment:
I had a go at this, but it's complicated by the fact that you have to return the result of the `operation`, plus `Repeater` doesn't have an exact equivalent of `isExceptionRetryable`, so you have to capture the exception in the `repeat` body and save it for the `until` to look at.  It's doable but didn't seem compellingly simpler than what's there now.